Generic KV Client Instrumentation

jtlisi opened the pull request:
**What this PR does**:

- Add a name and backend type label to the prometheus metrics that wrap KV clients
- Remove dedicated instrumentation for the consul client in favor of a generic instrumentation for all clients

**Which issue(s) this PR fixes**:
Fixes #2484 

**Checklist**
- [x] Tests updated
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented on pkg/ring/kv/client.go:
```diff
@@ -97,19 +99,19 @@ type Client interface {
 // NewClient creates a new Client (consul, etcd or inmemory) based on the config,
 // encodes and decodes data for storage using the codec.
-func NewClient(cfg Config, codec codec.Codec) (Client, error) {
+func NewClient(name string, cfg Config, codec codec.Codec, reg prometheus.Registerer) (Client, error) {
```
I think passing `name` to components that is used just for metrics registration is an anti-pattern. Caller of this call can easily wrap `reg` into `prometheus.WrapRegistererWith(prometheus.Labels{"name": "some name here"}, reg)` if it wishes to do so. I would prefer to use this pattern instead. KV store client doesn't need a name for anything else, and it also doesn't need it if registrerer is `nil`.

pstibrany commented on pkg/ring/kv/client.go:
```diff
@@ -149,10 +151,10 @@ func createClient(name string, prefix string, cfg StoreConfig, codec codec.Codec
 	}
 
-	return metrics{client}, nil
+	return newMetricsClient(name, backend, client, reg), nil
```
If `reg` is nil (which is a case we should support and is typical for tests), we can skip creation of metrics client completely.

jtlisi replied:
👍 Make sense

jtlisi replied:
Makes sense

pstibrany commented on pkg/ring/lifecycler.go:
```diff
@@ -147,7 +147,8 @@ func NewLifecycler(cfg LifecyclerConfig, flushTransferer FlushTransferer, ringNa
 	codec := GetCodec()
-	store, err := kv.NewClient(cfg.RingConfig.KVStore, codec)
+	// Suffix all client names with "-lifecycler" to denote this kv client is used by the lifecycler
+	store, err := kv.NewClient(ringName+"-lifecycler", cfg.RingConfig.KVStore, codec, nil)
```
This is a great opportunity to pass registrerer to lifecycler and move lifecycler metrics away from global variables.

pracucci replied:
But maybe in another PR?

jtlisi replied:
I will pass the Registerer but keep the lifecycler metrics global for now.

pstibrany commented on pkg/distributor/ha_tracker_test.go:
```diff
@@ -135,7 +136,7 @@ func TestWatchPrefixAssignment(t *testing.T) {
 		UpdateTimeoutJitterMax: 0,
 		FailoverTimeout:        time.Millisecond * 2,
-	})
+	}, prometheus.NewRegistry())
```
`nil`, unless metrics are checked.

jtlisi replied:
👍

pstibrany commented on pkg/ring/kv/consul/client.go:
```diff
@@ -85,7 +84,7 @@ func NewClient(cfg Config, codec codec.Codec) (*Client, error) {
 		return nil, err
 	}
 	c := &Client{
-		kv:    consulMetrics{client.KV()},
```
Previously we could see individual `CAS` calls to consul. Now we will only see "CAS loop".

pracucci commented:
> Better way of handling multiple components with same metrics is to wrap registrerer to add constant label to all metrics registered by component

Agree. It's more powerful.

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -184,7 +184,7 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 	cfg.PoolConfig.RemoteTimeout = cfg.RemoteTimeout
 
-	replicas, err := newClusterTracker(cfg.HATrackerConfig)
+	replicas, err := newClusterTracker(cfg.HATrackerConfig, nil)
```
Why don't we pass the registerer here?

jtlisi replied:
The distributor didn't pass a registerer. I added one.

pracucci commented on pkg/distributor/ha_tracker.go:
```diff
@@ -157,7 +157,7 @@ func newClusterTracker(cfg HATrackerConfig) (*haTracker, error) {
 
 	if cfg.EnableHATracker {
-		client, err := kv.NewClient(cfg.KVStore, GetReplicaDescCodec())
+		client, err := kv.NewClient(RingNameForClient, cfg.KVStore, GetReplicaDescCodec(), reg)
```
I think using the same name here would clash with the distributor ring name. We could just pass `distributor-hatracker` as ring client name. What do you think? /cc @pstibrany

jtlisi replied:
Makes sense

pracucci commented on pkg/ring/kv/metrics.go:
```diff
@@ -32,12 +22,36 @@ func errorCode(err error) string {
+}
+
+func newMetricsClient(name string, backend string, c Client, reg prometheus.Registerer) Client {
+	// TODO: Ensure all clients are passed a valid registry and remove the following check
+	if reg == nil {
+		reg = prometheus.DefaultRegisterer
+	}
```
Agree. This TODO should be removed (and the default register never used).

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -139,7 +139,7 @@ func NewBlocksStoreQueryableFromConfig(querierCfg Config, gatewayCfg storegatewa
 	if gatewayCfg.ShardingEnabled {
 		storesRingCfg := gatewayCfg.ShardingRing.ToRingConfig()
-		storesRingBackend, err := kv.NewClient(storesRingCfg.KVStore, ring.GetCodec())
+		storesRingBackend, err := kv.NewClient(storegateway.RingNameForClient+"-block-store", storesRingCfg.KVStore, ring.GetCodec(), reg)
```
The generated name here looks a bit weird to me. Maybe `querier-store-gateway` may be more clear?

jtlisi replied:
Sounds good, I wasn't sure how best to name some of these clients.

pracucci commented on pkg/ring/kv/consul/client.go:
```diff
@@ -99,21 +98,17 @@ func (c *Client) Put(ctx context.Context, key string, value interface{}) error {
 		return err
 	}
 
-	return instrument.CollectedRequest(ctx, "Put", consulRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
```
Now that I notice, I think previous version was even wrong. We did track "Put" twice for 1 Put call. Once here, and once in `consul/metrics.go`. Am I missing anything?

jtlisi replied:
After investigating further, the consul metrics probably should remain as is. They provide lower level metrics for the client interacting with consul itself: https://github.com/hashicorp/consul/blob/master/api/kv.go

In our KV implementation CAS is retried multiple times and by using a generic metrics client we are missing out on instrumenting some of these calls. One option that may be used to solve this problem is moving retries to there own KV implementation and wrap clients with it.

jtlisi replied:
I am going to re-add the separate consul instrumentation and will add an issue to add a generic retry implementation for the KV store interface. WDYT?

pracucci commented on pkg/ring/kv/metrics.go:
```diff
@@ -32,12 +22,36 @@ func errorCode(err error) string {
+				Name:      "kv_request_duration_seconds",
+				Help:      "Time spent on kv store requests.",
+				Buckets:   prometheus.DefBuckets,
+				ConstLabels: prometheus.Labels{
+					"name": name,
+					"type": backend,
+				},
```
Please mention this change in the CHANGELOG. You should also check if it has an impact on the cortex mixin's dashboards (depends on how we group series, I haven't checked).

pracucci commented on pkg/ring/kv/consul/metrics.go:
```diff
@@ -1,82 +0,0 @@
-	"github.com/weaveworks/common/instrument"
-)
-
-var consulRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
```
Please remember to mention in the CHANGELOG that the metric `cortex_consul_request_duration_seconds` has been removed.

jtlisi replied:
I re-added this metric since it fits a different use case.

jtlisi commented:
@pstibrany & @pracucci 

I updated this PR per your comments. Changes include:

- Keeping the `cortex_consul_request_duration_seconds` metric and consul instrumentation w/ note about a possible way of removing it in a future PR
- Registerers are now plumbed through to each call of the kv.NewClient
- Updated ChangeLog

I will test this PR w/ the cortex-mixin this week and ensure it operates correctly with the new metrics.

If you have anymore feedback about the naming of KV clients let me know. I wasn't sure if some of the names/suffixes I chose made sense.

pstibrany commented on pkg/distributor/distributor_test.go:
```diff
@@ -369,7 +369,7 @@ func TestDistributor_PushHAInstances(t *testing.T) {
 						UpdateTimeout:   100 * time.Millisecond,
 						FailoverTimeout: time.Second,
-					})
+					}, prometheus.NewRegistry())
```
```suggestion
					}, nil)
```

Since we don't verify metrics.

pstibrany commented on pkg/ring/kv/metrics.go:
```diff
@@ -32,12 +37,35 @@ func errorCode(err error) string {
+}
+
+func newMetricsClient(backend string, c Client, reg prometheus.Registerer) Client {
+	// If no Registerer is provided return the raw client
+	if reg == nil {
+		return c
+	}
```
Nit: I'd move this check into `createClient` function.

pstibrany commented on pkg/ring/kv/metrics.go:
```diff
@@ -5,19 +5,24 @@ import (
+		return nil
+	}
+
+	return prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
```
Is it worth using something more unique, e.g. `kv-name` here?

jtlisi replied:
That makes sense. I was just basing it off the label we use for the ring.

pstibrany commented on pkg/ring/kv/metrics.go:
```diff
@@ -5,19 +5,24 @@ import (
-	Buckets:   prometheus.DefBuckets,
-}, []string{"operation", "status_code"}))
+var (
+	primaryLabel   = prometheus.Labels{"role": "primary"}
+	secondaryLabel = prometheus.Labels{"role": "secondary"}
```
Only used from client.go, I would move it into that file.

pstibrany commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## master / unreleased
 
+* [CHANGE] Metric `cortex_kv_request_duration_seconds` now includes `name` label to denote which client is being used as well as the `backend` label to denote the KV backend implemnetation in use. #2648
```
```suggestion
* [CHANGE] Metric `cortex_kv_request_duration_seconds` now includes `name` label to denote which client is being used as well as the `backend` label to denote the KV backend implementation in use. #2648
```

pstibrany commented on pkg/distributor/distributor_ring.go:
```diff
@@ -13,6 +13,10 @@ import (
 	"github.com/cortexproject/cortex/pkg/util/flagext"
 )
 
+var (
+	RingNameForClient = "distributor"
+)
+
```
```suggestion
const (
	ringNameForClient = "distributor"
)
```

And move to `distributor_ring.go` where it is used.

(Alternatively, just inline it)

pstibrany commented on pkg/ruler/ruler.go:
```diff
@@ -190,7 +190,11 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable promStorage.Queryable
+		ringStore, err := kv.NewClient(
+			cfg.Ring.KVStore,
+			ring.GetCodec(),
+			kv.RegistererWithKVName(reg, ring.RulerRingKey),
```
```suggestion
			kv.RegistererWithKVName(reg, "ruler"),
```

`ring.RulerRingKey` is actually `ring`.

pstibrany commented on pkg/storegateway/gateway.go:
```diff
@@ -81,7 +81,11 @@ func NewStoreGateway(gatewayCfg Config, storageCfg cortex_tsdb.Config, logLevel
+		ringStore, err = kv.NewClient(
+			gatewayCfg.ShardingRing.KVStore,
+			ring.GetCodec(),
+			kv.RegistererWithKVName(reg, RingNameForClient),
```
```suggestion
			kv.RegistererWithKVName(reg, "store-gateway"),
```
